### ISSUE TITLE
Create a mechanism to benchmark GPU mask mode.

### ISSUE THREAD

**Issue by claudioandre** [magnumripper/JohnTheRipper#1822]
Since it could be hard to implement a full/regular benchmark for GPU mask mode, I created a hack that might help:
* It can be disabled if desired.
* It does not impact a regular run.
* Even when enabled, it does not interfere with a normal run.

Please, take a look and check if this is desirable, useful and mergeable. Patch, 
ec2aa6ccd9ebfcf401bb7caa81a9bd9c8429e672

To use it (skip_self_test is mandatory)
```_MASK_MODE_DSV0=1 ../run/john -test -form:raw-sha256-opencl  -dev=0 --skip ```

**Comment by magnumripper** [comment]
Maybe instead of using that environment variable, we could let `-test -mask=(...)` imply this (and also imply -skip-self-test). I tried something similar a while ago but got stuck with options.c not allowing -test and -mask simultaneously and for some reason I couldn't get past that.

That would also mean we can choose any mask we want for benchmarking.

**Comment by claudioandre** [comment]
Since you got stuck in core code, I will wait to see if you get a way to solve it.

**Comment by magnumripper** [comment]
Should there be any need for per-format code for this? My approach was to simply try to run a hybrid mask-mode code path (and ignore self-test).

**Comment by claudioandre** [comment]
> Should there be any need for per-format code for this?

No. Your approach is much better.

**Comment by magnumripper** [comment]
@claudioandre @Sayantan2048 I think I did it... some testing remains.
```
$ ../run/john -test -form:nt-opencl -dev=6 
Device 6: GeForce GTX TITAN X
Benchmarking: nt-opencl, NT [MD4 OpenCL]... DONE
Raw:	62128K c/s real, 62128K c/s virtual

$ ../run/john -test -form:nt-opencl -dev=6 -mask
Device 6: GeForce GTX TITAN X
Benchmarking: nt-opencl, NT [MD4 OpenCL]... DONE
Raw:	11193M c/s real, 11193M c/s virtual
```

You can also specify a mask (including hybrid ones) to see performance for eg. a very short mask.
```
$ ../run/john -test -form:nt-opencl -dev=6 -mask=?w?a?a?a
Device 6: GeForce GTX TITAN X
Benchmarking: nt-opencl, NT [MD4 OpenCL]... DONE
Raw:	11419M c/s real, 11533M c/s virtual

$ ../run/john -test -form:nt-opencl -dev=6 -mask=?a?a?a?w
Device 6: GeForce GTX TITAN X
Benchmarking: nt-opencl, NT [MD4 OpenCL]... DONE
Raw:	11311M c/s real, 11199M c/s virtual

$ ../run/john -test -form:nt-opencl -dev=6 -mask=?d      
Device 6: GeForce GTX TITAN X
Benchmarking: nt-opencl, NT [MD4 OpenCL]... DONE
Raw:	677421K c/s real, 677421K c/s virtual
```

We might have some tweaking to do. I'm not sure whether all formats actually auto-tune correctly for this feature. NT seems to do but some salted ones might not. But this should be a good start!

**Comment by magnumripper** [comment]
DEScrypt does not work well with this
```
$ ../run/john -test -dev=6 -form:descrypt-opencl 
Device 6: GeForce GTX TITAN X
Benchmarking: descrypt-opencl, traditional crypt(3) [DES OpenCL]... Salt compiled from Source:910
Salt compiled from Binary:910
Salt compiled from Binary:2275
Salt compiled from Binary:990
Salt compiled from Binary:0
DONE
Warning: "Many salts" test limited: 2/256
Many salts:	134217K c/s real, 132888K c/s virtual
Only one salt:	61567K c/s real, 61008K c/s virtual

$ ../run/john -test -dev=6 -form:descrypt-opencl -mask 
Device 6: GeForce GTX TITAN X
Benchmarking: descrypt-opencl, traditional crypt(3) [DES OpenCL]... Salt compiled from Source:910
Salt compiled from Binary:910
Salt compiled from Binary:2275
Salt compiled from Binary:990
Salt compiled from Binary:0
DONE
Warning: "Many salts" test limited: 2/256
Many salts:	7489K c/s real, 7423K c/s virtual
Only one salt:	7049K c/s real, 7049K c/s virtual
```

LM does get faster, but not optimal speed. I guess both of them currently doesn't auto-tune for mask when this feature is used - some tweaks will be needed.
```
$ ../run/john -test -dev=6 -form:lm-opencl 
Device 6: GeForce GTX TITAN X
Benchmarking: LM-opencl [DES BS OpenCL]... DONE
Raw:	96559K c/s real, 95869K c/s virtual

$ ../run/john -test -dev=6 -form:lm-opencl -mask 
Device 6: GeForce GTX TITAN X
Benchmarking: LM-opencl [DES BS OpenCL]... DONE
Raw:	765671K c/s real, 765671K c/s virtual
```

**Comment by claudioandre** [comment]
Kudos.

Please, commit in a branch so we can test it.

**Comment by magnumripper** [comment]
It is committed, 4592cc6b

**Comment by magnumripper** [comment]
Right now I'm actually never calling `do_mask_crack(word)` from `benchmark_format` or anything like that - I just init mask mode. Maybe this is not quite sufficient.

**Comment by magnumripper** [comment]
> I guess both of them currently doesn't auto-tune for mask when this feature is used

They do. It's probably the lack of `do_mask_crack()` or `generate_keys()` or whatever.

**Comment by claudioandre** [comment]
Seems to be fine. I will change some details in format code and submit. 
* Is 500ms is not good enough? No.
* Remove: self->params.max_keys_per_crypt /= 256;
* Etc

```
		//Auto tune execution from shared/included code.
		autotune_run(self, 1, gws_limit, 500ULL);

		load_hash(NULL);

		if (options.flags & FLG_MASK_CHK) {
			fprintf(stdout,
			        "Using Mask Mode with internal candidate generation%s",
			        flag ? "" : "\n");

			if (flag) {
				self->params.max_keys_per_crypt /= 256;

				if (self->params.max_keys_per_crypt < 1)
					self->params.max_keys_per_crypt = 1;
```

**Comment by magnumripper** [comment]
500ms is plenty. I think Sayantan aims for 300ms in some format I visited.

**Comment by magnumripper** [comment]
Hold on, I will commit enhancements.

**Comment by magnumripper** [comment]
c8d543c added more proper handling. We now do call `do_mask_crack()` although it's handled a bit different under --test.

DEScrypt and LM formats has a weird bug but it appears this is an old bug (using default mask):
```
$ ../run/john -dev=6 -test -form:lm-opencl -mask

Make process completed.
Device 6: GeForce GTX TITAN X
Benchmarking: LM-opencl [DES BS OpenCL]... Format internal ranges cannot be truncated!
Use a bigger key length or non-gpu format.
```

**Comment by claudioandre** [comment]
> 500ms is plenty.

I get your point, but remember, I'm always thinking about super --dev=0.  Example:
- raw-md5-opencl is using 300ms. You can see it picks GWS=1024 on -dev=0. If you use GWS=2048 the speed is 2x better.

**Comment by claudioandre** [comment]
Forget it. The solution is more brain and less bla bla bla.

**Comment by magnumripper** [comment]
I also see now 300ms is a bit on the low side for weak GPUs (like my laptop's GT650). But instead of increasing auto-tune time, it should be better to decrease `mask_int_cand_target` for weak devices.

Ideally we should have a simply and fast shared function you could call for a rough "performance class" figure for a device. But what should it return? Perhaps a duration, in ms, for a given speed-test kernel?

**Comment by claudioandre** [comment]
Indeed. I reduced ```mask_int_cand_target```. Have to see if/how it will impact strong GPUs.

**Comment by magnumripper** [comment]
It's probably hard to find a good compromise. A Titan X can do over 15-20 Gp/s with NT. My laptop can do 0.3 Gp/s.

Also, the lower number for mask_int_cand_target, the less chance for mask mode to find a good multiple (eg. if 95x95 is too high, it has to decrease to just 95, which is too low for a good GPU).

**Comment by magnumripper** [comment]
That "first GWS check in half target time" that Solar added was one of the problems. Fixed in ce8c882.

**Comment by magnumripper** [comment]
I see you had it has 20000 though. You should definitely decrease it to 10000. No other format has over 10000, DEScrypt has just 1024.

**Comment by magnumripper** [comment]
That problem solved canonically in #1831 :sunglasses:
